apps/api/pages/api/destination-calendars/[id]/_patch.ts

```diff
@@ -57,15 +60,131 @@ import { schemaQueryIdParseInt } from "~/lib/validations/shared/queryIdTransform
+  }
+
+  if (parsedBody.integration && parsedBody.externalId) {
+    const calendarCredentials = getCalendarCredentials(userCredentials);
+
+    const { connectedCalendars } = await getConnectedCalendars(
+      calendarCredentials,
+      [],
+      parsedBody.externalId
+    );
+    const connectedCalendar = connectedCalendars.find(
+      (c) =>
+        c?.primary?.externalId === parsedBody.externalId && c?.primary?.integration === parsedBody.integration
+    );
+
+    if (!connectedCalendar)
+      throw new HttpError({
+        statusCode: 400,
+        message: "Bad request, credential id invalid",
+      });
+
+    credentialId = connectedCalendar.primary?.credentialId;
+  }
 
   const destinationCalendar = await prisma.destinationCalendar.update({
     where: { id },
-    data: parsedBody,
+    data: { ...parsedBody, credentialId },
   });
   return { destinationCalendar: schemaDestinationCalendarReadPublic.parse(destinationCalendar) };
 }
 
+async function findUserCredentials(req: NextApiRequest) {
+  const { userId: requestUserId, prisma, isAdmin } = req;
+  const { id } = schemaQueryIdParseInt.parse(req.query);
+  const body = schemaDestinationCalendarEditBodyParams.parse(req.body);
+  const userId = isAdmin ? body.userId || requestUserId : requestUserId;
+  const destinationCalendarObject = await prisma.destinationCalendar.findFirst({
+    where: {
+      id,
+    },
+    select: { eventTypeId: true, credentialId: true },
+  });
+
+  if (!destinationCalendarObject) {
+    throw new HttpError({
+      statusCode: 404,
+      message: `Destination calendar with ID ${id} not found`,
+    });
+  }
+
+  if (!destinationCalendarObject.credentialId) {
+    throw new HttpError({
+      statusCode: 404,
+      message: `Destination calendar missing credential id`,
+    });
+  }
+
+  const credentials = await prisma.credential.findMany({
+    where: { id: destinationCalendarObject.credentialId, userId },
+    select: credentialForCalendarServiceSelect,
+  });
+
+  if (!credentials || credentials.length === 0) {
+    throw new HttpError({
+      statusCode: 400,
+      message: `Bad request, no associated credentials found`,
+    });
+  }
+
+  if (body.eventTypeId) {
+    if (destinationCalendarObject.eventTypeId) {
+      const userEventType = await prisma.eventType.findFirst({
+        where: { id: body.eventTypeId },
+        select: { userId: true },
+      });
+
+      if (!userEventType || userEventType.userId !== userId) {
+        throw new HttpError({
+          statusCode: 404,
+          message: `Event type with ID ${body.eventTypeId} not found`,
+        });
+      }
+      return { userCredentials: credentials };
+    }
+    throw new HttpError({
+      statusCode: 400,
+      message: `The provided destination calendar can not be linked to an event type`,
+    });
+  }
+
+  if (!body.eventTypeId) {
+    if (!destinationCalendarObject.eventTypeId) return { userCredentials: credentials };
```

Comment:
this seems like checks that could be done before trying to get the userCredentials

Reply:
let me know if this makes sense, we can look together if needed